File: Min Stack/Min_Stack.cpp
// 利用链表实现一个最小堆栈
// 思路十分巧妙

// Runtime: 40 ms, faster than 92.40% of C++ online submissions for Min Stack.
// Memory Usage: 17.1 MB, less than 100.00% of C++ online submissions for Min Stack.

struct linkNode
{
    int val;
    linkNode *next;
    linkNode(int value)
    {
        val = value;
        next = 0;
    }
};

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() 
    {
        head = new linkNode(0);
        minvalue = INT_MAX;
    }
    
    void push(int x) 
    {
        if (x <= minvalue)
        {
            linkNode *tempNode = new linkNode(minvalue);
            if (empty())
                head->next = tempNode;
            else
            {
                tempNode->next = head->next;
                head->next = tempNode;
            }
            minvalue = x;
        }

        linkNode *tempNode = new linkNode(x);
        if (empty())
            head->next = tempNode;
        else
        {
            tempNode->next = head->next;
            head->next = tempNode;
        }
    }
    
    void pop() 
    {
        if (!empty())
        {
            if (head->next->val == minvalue)
            {
                linkNode *deleteNode = head->next;
                head->next = deleteNode->next;
                delete deleteNode;
                minvalue = head->next->val;
            }

            linkNode *deleteNode = head->next;
            head->next = deleteNode->next;
            delete deleteNode;
        }    
    }
    
    int top() 
    {
        return head->next->val;
    }
    
    int getMin() 
    {
        return minvalue;
    }

private:
    linkNode *head;
    int minvalue;

    bool empty()
    {
        if (head->next == 0)
            return true;
        else 
            return false;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Min Stack/Min_Stack.py
# 注意一下python中是如何实现一个链表的，比c++要容易很多

# Runtime: 56 ms, faster than 91.81% of Python3 online submissions for Min Stack.
# Memory Usage: 17 MB, less than 100.00% of Python3 online submissions for Min Stack.

class linkNode:
    def __init__(self, value):
        self.val = value
        self.next = None

class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.head = linkNode(0)
        self.minvalue = float('inf')

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x <= self.minvalue:
            tempNode = linkNode(self.minvalue)
            if self.empty():
                self.head.next = tempNode
            else:
                tempNode.next = self.head.next
                self.head.next = tempNode
            self.minvalue = x
        
        tempNode = linkNode(x)
        if self.empty():
            self.head.next = tempNode
        else:
            tempNode.next = self.head.next
            self.head.next = tempNode
        
    def pop(self):
        """
        :rtype: void
        """
        if not self.empty():
            if self.head.next.val == self.minvalue:
                deleteNode = self.head.next
                self.head.next = deleteNode.next
                self.minvalue = self.head.next.val
            deleteNode = self.head.next
            self.head.next = deleteNode.next
        
    def top(self):
        """
        :rtype: int
        """
        return self.head.next.val
        
    def getMin(self):
        """
        :rtype: int
        """
        return self.minvalue
    
    def empty(self):
        if self.head.next is None:
            return True
        else:
            return False

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()